### DISCUSSION TITLE
display 为 none 的时候，为啥不会删值

### DISCUSSION THREAD

**Discussion by javahuang** [alibaba/formily#1665] [Q&A 2.x] [answered]
示例：[codesandbox](https://codesandbox.io/s/flamboyant-tu-jmhg1?file=/App.tsx)

- aa(VoidField)
  -  bb(Field)
  - cc(Field)

复现步骤：

1. bb 输入 1，cc 输入 123
2. 点击 hide

此时，cc 的值为 123，但是 display 为 none。

我看源码里面，aa(VoidField) 的 display 为 none 时，由于 bb 没有设置 selfDisplay，此时 bb 的 display 发生了改变，会触发 reaction 执行删值；
接着会触发 cc 字段的配置的 reactions，然后会设置 cc 的 display 为 none，但是为啥没有接着触发 cc Field reaction 的删值行为呢？

**Comment by janryWang** [comment, marked as answer]
因为cc被设置了selfDisplay，父节点display变化对它自身不影响的，所以Void父节点display none，它自身的数据并不会删除，子节点的数据也不会被删除，所以行为是合理的

**Reply by javahuang** [reply]
设置 cc 的 display = none，display compute为啥不会触发这段代码执行呢。我给示例里面的注释放开，就会删值，不太明白这里面 reaction 触发的原理，求解惑。

https://github.com/alibaba/formily/blob/acec46fc4e84be5c97ed036f7d286312417386f4/packages/core/src/models/Field.ts#L243-L258

我一直是对比着这 mobx 来学习 formily/reactive的，我看 mobx 里面有这条规则

> Reactions shouldn't update other observables: 

但是在 formily 里面，很多地方是利用了 reactions 来改变 observable 再次触发 reaction 的。

**Reply by janryWang** [reply]
设置cc的display的过程其实是在设置selfDisplay，你看看get display的逻辑就知道了，相当于响应被selfDisplay拦截了

**Reply by janryWang** [reply]
记住一点就行，当前节点的display优先级高于父节点优先级

**Reply by javahuang** [reply]
1.  aa 字段设置 display=none
2.  bb 字段的 parentDisplay 发生了变化，导致 bb 字段的 display 重新计算，此时 bb 字段的 display = none，触发了 bb Field 内部的 visible reaction 触发删值行为
3. 在 cc 字段设置的 reactions 里面依赖了 bb 字段的值，导致该 reaction 触发
4. 在 cc 字段的 reaction 里面设置 cc 的 display=none
5. 此时 cc 的 selfDisplay=none，display 会重新计算，但为啥不会继续触发 cc Field 内部的 visible reaction 呢？

**Comment by javahuang** [comment]
## `observable.computed` 的触发时机

```ts
import {
  define,
  observable,
  reaction,
} from "@formily/reactive";

export class Javahuang {
  name: string;
  age: number;

  constructor() {
    this.name = "javahuang";
    this.age = 30;

    define(this, {
      name: observable.ref,
      age: observable.ref,
      salary: observable.computed,
    });

    reaction(
      () => this.salary,
      (salary) => {
        console.log(`javahuang's salary is ${salary}`);
      }
    );
  }

  get salary() {
    if (this.name === "javahuang") {
      return 100;
    }
    return this.age * 10;
  }
}

const huang  = new Javahuang();
huang.age = 300; // 不会触发
huang.name = 'dahuang'; // 会触发
huang.name = 'javahuang'; // 继续触发
huang.age = 20; // 不会触发
```

昨天 @janryWang 给我科普了一下 compute，在这个地方我再记录一下。

compute 依赖里面的 observable 属性，如果 observable 的值发生了变化，会触发 compute 属性的重新计算。
每次 compute 计算会重新收集一遍依赖，遇到 if 语句就会停止收集，只有收集的 observable 发生变化才会触发再次计算。
另外，对于 compute 属性绑定的 reaction，是对 compute 的值做了浅比较，依据比较结果看是否触发对应的 handler。

如上面的示例：

1. Javahuang 实例化之后会初始计算一遍 salary
2. 此时由于 `this.name === "javahuang"`，所以会停止收集，也就是说如果只改变 age 是不会触发 salary 的重新计算的
3. 我们接着设置 `huang.name = 'dahuang'`，此时会触发 compute salary，会收集 name 和 age，后面无论是改变 name 还是 age 都会触发 compute salary

**Comment by yuyi919** [comment]
最后这个问题是怎么解决的？看主楼里的demo还是没解决

**Reply by janryWang** [reply]
已经修复了，等下个版本

**Reply by febugcoder** [reply]
> 已经修复了，等下个版本

v2.0.0-beta.73好像还是有问题，官网的例子：
https://codesandbox.io/s/77si0
input填值，选择none，再选择visable，input的值还在

**Reply by janryWang** [reply]
@febugcoder 你完全不理解这个bug是啥bug。。。你这个是合理行为

**Reply by febugcoder** [reply]
> @febugcoder 你完全不理解这个bug是啥bug。。。你这个是合理行为

仔细再看了一下，确实是我理解的问题，现在是修复了
